nrpy/infrastructures/BHaH/main_c.py

```diff
@@ -96,43 +131,88 @@ def register_CFunction_main_c(
 //           Then overwrite the parameter file values with those provided via command line arguments.
 cmdline_input_and_parfile_parser(&commondata, argc, argv);
-
+"""
+    griddata_malloc = r"""
 // Step 1.c: Allocate memory for MAXNUMGRIDS griddata structs,
 //           where each structure contains data specific to an individual grid.
 griddata = (griddata_struct *restrict)malloc(sizeof(griddata_struct)*MAXNUMGRIDS);
+"""
+
+    if parallelization in ["cuda"]:
+        griddata_malloc = griddata_malloc.replace("griddata = ", "griddata_device = ")
+        griddata_malloc += "griddata_host = (griddata_struct *)malloc(sizeof(griddata_struct) * MAXNUMGRIDS);\n"
 
+    body += griddata_malloc
+    body += r"""
 // Step 1.d: Initialize each CodeParameter in griddata.params to its default value.
 params_struct_set_to_default(&commondata, griddata);
+""".replace(
+        "griddata", "griddata_device" if parallelization in ["cuda"] else "griddata"
+    )
 
+    body += r"""
 // Step 1.e: Set up numerical grids, including parameters such as NUMGRIDS, xx[3], masks, Nxx, dxx, invdxx,
 //           bcstruct, rfm_precompute, timestep, and others.
 {
   // If this function is being called for the first time, initialize commondata.time, nn, t_0, and nn_0 to 0.
   const bool calling_for_first_time = true;
   numerical_grids_and_timestep(&commondata, griddata, calling_for_first_time);
 }
-
+""".replace(
+        "griddata,",
+        (
+            "griddata_device, griddata_host,"
+            if parallelization in ["cuda"]
+            else "griddata,"
+        ),
+    )
+    body += r"""
 for(int grid=0; grid<commondata.NUMGRIDS; grid++) {
   // Step 2: Allocate storage for the initial data (y_n_gfs gridfunctions) on each grid.
   MoL_malloc_y_n_gfs(&commondata, &griddata[grid].params, &griddata[grid].gridfuncs);
 }
+""".replace(
+        "griddata", "griddata_device" if parallelization in ["cuda"] else "griddata"
+    )
+    if parallelization == "cuda":
+        body += r"""
+for(int grid=0; grid<commondata.NUMGRIDS; grid++) {
+  // Step 2.b: Allocate host storage for diagnostics
+  CUDA__malloc_host_gfs(&commondata, &griddata_device[grid].params, &griddata_host[grid].gridfuncs);
+  CUDA__malloc_host_diagnostic_gfs(&commondata, &griddata_device[grid].params, &griddata_host[grid].gridfuncs);
+}
 """
-    setup_initial_data_code = """Set up initial data.
+    setup_initial_data_code = (
+        """Set up initial data.
+initial_data(&commondata, griddata_host, griddata_device);
+"""
+        if parallelization in ["cuda"]
+        else """Set up initial data.
 initial_data(&commondata, griddata);
 """
+    )
     allocate_storage_code = """Allocate storage for non-y_n gridfunctions, needed for the Runge-Kutta-like timestepping.
 for(int grid=0; grid<commondata.NUMGRIDS; grid++)
   MoL_malloc_non_y_n_gfs(&commondata, &griddata[grid].params, &griddata[grid].gridfuncs);
-"""
+""".replace(
```

Comment:
Let's remove all the replace() stuff and instead set the variable 
`griddata_dev = "griddata_device" if parallelization in ["cuda"] else "griddata"`
at the top and use f-strings.